### PR TITLE
Task3

### PR DESCRIPTION
**Pull request by echekholskiy**

### CODE REVIEW

File: VKApp/Extensions/AuthorizePageExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+using JDI.Light;
+using System.Text.RegularExpressions;
+using Forms.Pages;
+using Forms.Utils;
+
+namespace Forms.Extensions
+{
+    public static class AuthorizePageExtensions
+    {
+        public static void OpenAuthorizePageInNewWindowAndSwitch(this AuthorizePage authorizePage)
```
Comment (asushkov-github) [inline VKApp/Extensions/AuthorizePageExtensions.cs:11]: Why do you need extension, if you don't use authorizePage variable? Let's move it to AutenticationUtils class

File: VKApp/Pages/MainPage.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        public void LikePost(string ownerId, int postId)
+        {
+            Jdi.WebDriver.FindElement(By
+                .XPath($"//div[@id='post{ownerId}_{postId}']//*[@class='like_button_icon']"))
+                .Click();
+        }
+
+        public bool PostIsDeleted(string ownerId, int postId)
+        {
+            return new TextField(By.XPath($"//div[@id='wpt{ownerId}" + "_" + $"{postId}']")).Displayed;
```
Comment (asushkov-github) [inline VKApp/Pages/MainPage.cs:34]: You don't need '+' in interpolation

File: VKApp/Utils/ApiClient.cs [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+using Polly;
+using Polly.Retry;
+using RestSharp;
+
+namespace Forms.Utils
+{
+    public class ApiClient
+    {
+        private const int RetryCount = 5;
+        private const int Timeout = 2;
```
Comment (asushkov-github) [inline VKApp/Utils/ApiClient.cs:18]: TimeoutInSeconds

File: VKApp/Utils/ApiClient.cs [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+            .WaitAndRetry(RetryCount, retryAttempt =>
+                TimeSpan.FromSeconds(Math.Pow(Timeout, retryAttempt)));
+        private readonly string _apiVersion = Configuration.ApiVersion;
+        public readonly AuthorizePage AuthorizePage = new();
+        private readonly string _token;
+        public static ApiClient Instance => LazyInstance.Value;
+
+        private ApiClient()
+        {
+            _token = GetToken();
```
Comment (asushkov-github) [inline VKApp/Utils/ApiClient.cs:34]: `_token = AuthenticationUtils.GetToken()`

File: VKApp/Utils/ApiClient.cs [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+            Jdi.Logger.Info("Upload photo on server by API");
+
+            var rest = new RestClient(url);
+            var request = new RestRequest(Method.POST)
+            {
+                AlwaysMultipartFormData = true
+            };
+            request.AddHeader(ApiParameterKeys.ContentType, ApiParameters.Multipart);
+            request.AddFile(ApiParameterKeys.File,
+                File.ReadAllBytes(FileManager.GetImagePath()),
```
Comment (asushkov-github) [inline VKApp/Utils/ApiClient.cs:80]: `FileManager.GetImagePath().GetBytes()`

File: VKApp/Utils/ApiClient.cs [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+                Path.GetFileName(Configuration.PictureName),
```
Comment (asushkov-github) [inline VKApp/Utils/ApiClient.cs:81]: `Configuration.PictureName.GetFileName()`

File: VKApp/Utils/ApiClient.cs [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+using Forms.Extensions;
+using Forms.Pages;
+using JDI.Light;
+using Polly;
+using Polly.Retry;
+using RestSharp;
+
+namespace Forms.Utils
+{
+    public class ApiClient
```
Comment (asushkov-github) [inline VKApp/Utils/ApiClient.cs:15]: I would suggest you to split this class on ApiClient base abstract class and VkApiClient implementation.

**Review by asushkov-github — CHANGES_REQUESTED** [review]

File: Task3/Test/Hooks.cs [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    [SetUpFixture]
+    public class Hooks
+    {
+        [OneTimeSetUp]
+        public void Initial()
+        {
+            Jdi.Init();
+            Activator.CreateInstance<LoginPage>();
+            Activator.CreateInstance<FeedPage>();
+            Activator.CreateInstance<MainPage>();
```
[the remark this answers is not in the document]
Comment (echekholskiy) [reply Task3/Test/Hooks.cs:18]: hmm, done

File: Task3/Test/TestContext.cs [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+
+            return this;
+        }
+
+        protected TestContext ThenLoginPageIsOpened()
+        {
+            var loginPageIsOpened = TestSite.LoginPage.IsOpened;
+            loginPageIsOpened
+                .Should()
+                .BeTrue("LoginPage is not opened");
```
[the remark this answers is not in the document]
Comment (echekholskiy) [reply Task3/Test/TestContext.cs:29]: done

File: Task3/Test/TestContext.cs [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+            mainPageIsOpened
+                .Should()
+                .BeTrue("MainPage is not opened");
+
+            return this;
+        }
+
+        protected TestContext WhenCreatedPostByApi()
+        {
+            _messageText = Faker.Country.Name();
```
[the remark this answers is not in the document]
Comment (echekholskiy) [reply Task3/Test/TestContext.cs:86]: done by Faker.Lorem.Paragraph(), ok?

File: Task3/Test/TestContext.cs [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+        protected TestContext WhenDeleteCreatedPost()
+        {
+            ApiClient.Instance.DeletePost(OwnerId, _postId);
+
+            return this;
+        }
+
+        protected TestContext ThenPostIsDeleted()
+        {
+            TestSite.MainPage.PostIsDeleted(OwnerId, _postId);
```
[the remark this answers is not in the document]
Comment (echekholskiy) [reply Task3/Test/TestContext.cs:177]: done

File: VKApp/Extensions/ApiClientExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+{
+    public static class ApiClientExtensions
+    {
+        public static string GetAttachment(this ApiClient apiClient, string ownerId)
+        {
+            var serverAddress = apiClient.GetServerAddress(ownerId);
+            var uploadOnServer = apiClient.UploadOnServer(serverAddress);
+            var savePhoto = apiClient.SavePhoto(ownerId, uploadOnServer);
+
+            return "photo" + savePhoto.Response[0].OwnerId + "_" + savePhoto.Response[0].Id;
```
[the remark this answers is not in the document]
Comment (echekholskiy) [reply VKApp/Extensions/ApiClientExtensions.cs:13]: done

File: VKApp/Pages/MainPage.cs (hunk 2 above)
Comment (echekholskiy) [reply VKApp/Pages/MainPage.cs:34]: done

File: VKApp/Utils/ApiClient.cs (hunk 3 above)
Comment (echekholskiy) [reply VKApp/Utils/ApiClient.cs:18]: done

File: VKApp/Utils/ApiClient.cs (hunk 5 above)
Comment (echekholskiy) [reply VKApp/Utils/ApiClient.cs:80]: by Extension? Or create GetImage method in FileManager?

File: Task3/Test/Hooks.cs [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+        }
+
+        [OneTimeTearDown]
+        public void AfterTest()
+        {
+            Jdi.WebDriver.Quit();
+            ImageManager.DeleteTestImage();
```
[the remark this answers is not in the document]
Comment (echekholskiy) [reply Task3/Test/Hooks.cs:25]: It's dowloaded test image.

File: VKApp/Utils/ApiClient.cs (hunk 4 above)
Comment (echekholskiy) [reply VKApp/Utils/ApiClient.cs:34]: done

File: VKApp/Extensions/AuthorizePageExtensions.cs (hunk 1 above)
Comment (echekholskiy) [reply VKApp/Extensions/AuthorizePageExtensions.cs:11]: done

File: Task3/Test/Test.cs [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+            WhenSuccessfulAuthorize();
+            ThenUserLoggedIn()
+                .And()
+                .FeedPageIsOpened();
+
+            WhenGoToMyPage();
+            ThenMainPageIsOpened();
+
+            WhenCreatedPostByApi();
```
[the remark this answers is not in the document]
Comment (echekholskiy) [reply Task3/Test/Test.cs:21]: done

File: Task3/Test/TestContext.cs [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+            postText
+                .Should()
+                .Be(_messageText, "Post text is not equals");
+
+            return this;
+        }
+
+        protected TestContext WhenEditPostWithPicture()
+        {
+            _messageText = Faker.Company.CatchPhrase();
```
[the remark this answers is not in the document]
Comment (echekholskiy) [reply Task3/Test/TestContext.cs:109]: -//-

File: VKApp/Utils/ApiClient.cs (hunk 7 above)
Comment (echekholskiy) [reply VKApp/Utils/ApiClient.cs]: done

File: Task3/Resources/appsettings.json [hunk 16]
```diff
@@ -0,0 +1,10 @@
+{
+  "ApiVersion": "5.92",
+  "PictureName": "picture.jpg",
+  "TestPictureName": "test.jpg",
+  "BaseUrl": "https://api.vk.com/method/",
+  "OwnerId": "657248114",
+  "Login": "+375336646663",
+  "Password": "Task31q2w3e",
+  "AuthenticationUrl": "https://oauth.vk.com/authorize?client_id=7869182&scope=204804&redirect_uri=https://oauth.vk.com/blank.html&display=page&response_type=token&revoke=1"
```
[the remark this answers is not in the document]
Comment (echekholskiy) [reply Task3/Resources/appsettings.json]: done

File: Task3/Test/TestContext.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (echekholskiy) [reply Task3/Test/TestContext.cs]: done

File: Task3/Test/TestContext.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (echekholskiy) [reply Task3/Test/TestContext.cs]: done

File: VKApp/Utils/ApiClient.cs (hunk 5 above)
Comment (echekholskiy) [reply VKApp/Utils/ApiClient.cs]: done

File: VKApp/Utils/ApiClient.cs (hunk 6 above)
Comment (echekholskiy) [reply VKApp/Utils/ApiClient.cs]: done

File: Task3/Test/TestContext.cs [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+            var attachments = ApiClient.Instance.GetAttachment(OwnerId);
+            ApiClient.Instance.EditPost(OwnerId, _postId, _messageText, attachments);
+
+            return this;
+        }
+
+        protected TestContext ThenPostCorrectEdited()
+        {
+            var imageUrl = TestSite.MainPage.GetPostImageUrl(OwnerId, _postId);
+            ImageManager.DownloadImageByUrl(imageUrl, Configuration.TestPictureName);
```
[the remark this answers is not in the document]
Comment (echekholskiy) [reply Task3/Test/TestContext.cs]: done

File: Task3/Test/Hooks.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (echekholskiy) [reply Task3/Test/Hooks.cs]: done

**Pull request closed without merging**